### PR TITLE
[BUG] fix `temporal_train_test_split` for hierarchical and panel data in case where `fh` is not passed

### LINKED ISSUE #5107: [BUG] `temporal_train_test_split` does not work on panel datatypes with unequal length series.
**Describe the bug**

Also relates to #4968 

`temporal_train_test_split` wrongly split panel datatypes (splitting per unique unequal  time series). It could be that the split function does not support this type yet, If so, it should throw an error msg telling that it does not currently supported the datatypes.

**To Reproduce**

```python
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.utils._testing.panel import _make_panel

y = _make_panel(n_instances=2, n_timepoints=6)
# make first instance series shorter than the second
y.iloc[4:6] =None
y.dropna(inplace=True)

train_size, test_size = temporal_train_test_split(y,  test_size=2) 
# show shapes
print(
    f"""
    {y.shape=} 
    {train_size.shape=}, {test_size.shape=} #train size should be (6,1) 2+4=6
    """
)

# has also the same issue as #4968 below is the minimal example
train_fh, test_fh = temporal_train_test_split(y,  fh=[1,2]) 
# show shapes
print(
    f"""
    {y.shape=} 
    {train_fh.shape=}, {test_fh.shape=} #train size should be (6,1) and test (4,1)
   """
)
```

output
```
y.shape=(10, 1) 
train_size.shape=(4, 1), test_size.shape=(4, 1) #train size should be 6 (2+4)

y.shape=(10, 1) 
train_fh.shape=(8, 1), test_fh.shape=(2, 1) #train size should be (6,1) and test (4,1)
```

**Versions**
<details>

 main at [3cf69ed](https://github.com/sktime/sktime/commit/3cf69eddba315d6130b661ca5fe8e132e236aa47)

</details>

### PR DESCRIPTION
**Pull request by fkiraly**
This PR fixes #5107 by replacing the inner workings of `temporal_train_test_split` with a splitter, making the design both DRY and reliant on splitter boilerplate.

This is done as follows:

* addition of a new splitter `TemporalTrainTestSplitter`, which encapsulates the `temporal_train_test_split` logic
* internally, `TemporalTrainTestSplitter` is a re-implementation of the `temporal_train_test_split` logic as a splitter class, giving access to the `BaseSplitter` boilerplate.
* removed osuleaf example, as that was time series classification and got split across instances in the sample rather than time points. This would be inconsistent with the general splitter interface and the docstring specification.

The issue with the panel data is solved by recurring to the existing broadcasting boilerplate of `BaseSplitter`.

Depends on https://github.com/sktime/sktime/pull/5332 for testing.

### CODE REVIEW

**Comment by fkiraly** [comment]
not sure why this fails - perhaps undocumented part of contract?

File: sktime/split/expandinggreedy.py [hunk 1]
```diff
@@ -16,30 +17,37 @@
...
+        If int: the number of steps included in the test set of each fold.
+            Formally, steps are consecutive ``iloc`` indices.
+        If float: the proportion of steps included in the test set of each fold,
+            as a proportion of the total number of index values.
+            Cave: not the ``loc`` proportion between start and end.
     folds : int, default = 5
         The number of folds.
     step_length : int, optional
         The number of steps advanced for each fold. Defaults to `test_size`.
+    reverse : bool, default = False
```
Comment (hazrulakmal) [inline sktime/split/expandinggreedy.py:43]: just wondering, when would this reverse argument be useful? 

Isn't forecasting is about forecasting the future? why would test set be at the start of the data?

**Comment by hazrulakmal** [comment]
> not sure why this fails - perhaps undocumented part of contract?

are you referring to the fh argument? I think the original author may not follow the convention of writing the docstring but he/she did specify the fh's expected input type.

File: sktime/split/expandinggreedy.py (hunk 1 above)
Comment (fkiraly) [reply sktime/split/expandinggreedy.py:43]: it's for repurposing the splitter to slice from the start of the time series, it would notb e useful by itself.

**Comment by fkiraly** [comment]
> are you referring to the fh argument?

no, to the failure that seems to arise from sth numpy

File: sktime/split/expandinggreedy.py (hunk 1 above)
Comment (fkiraly) [reply sktime/split/expandinggreedy.py]: I removed this again, I think you are right, this was strange.

**Comment by fkiraly** [comment]
works now - re-requested review @hazrulakmal

File: sktime/split/temporal_train_test_split.py [hunk 2]
```diff
@@ -83,46 +80,155 @@ def temporal_train_test_split(
...
-            # Get split into test and train data for timeindex only
-            series = (ys,)
-            yret = _train_test_split(
-                *series,
-                shuffle=False,
-                stratify=None,
-                test_size=test_size,
-                train_size=train_size,
-            )
+        return y_train, y_test
```
Comment (yarnabrina) [inline sktime/split/temporal_train_test_split.py:98]: This probably does not need an else block, since if block contains a return already.

**Review by yarnabrina — COMMENTED** [review]
Asked two clarification questions/doubts.

Also, may be it's a silly question, I couldn't spot a change that looked to be specifically for hierarchical/panel data. Can you please point me to the line that handles that part? Does this branch resolves both #4968 and #5107?

**Comment by fkiraly** [comment]
> Also, may be it's a silly question, I couldn't spot a change that looked to be specifically for hierarchical/panel data. Can you please point me to the line that handles that part?

Hm, we should also perhaps add tests for this?

The part that solves this imo is as I claimed in the PR preamble: "The issue with the panel data is solved by recurring to the existing broadcasting boilerplate of `BaseSplitter`."

In more detail, `temporal_train_test_split` now calls `TemporalTrainTestSplitter`'s `split_series`, which is the boilerplate in `BaseSplitter` (not new, hence DRY, and takes care of the hierarchical case) that ends up calling `_split`. This includes a broadcasting loop for hierarchical/panel data, all in the boilerplate layer of `BaseSplitter`, hence solving the issue in #5107.

It does not resolve #4968, as this deals only with the case where `fh` is not passed, the code for the case where `fh` is passed remains unchanged.

File: sktime/split/temporal_train_test_split.py [hunk 3]
```diff
@@ -25,52 +29,45 @@ def temporal_train_test_split(
     train_size: Optional[float] = None,
     fh: Optional[FORECASTING_HORIZON_TYPES] = None,
 ) -> SPLIT_TYPE:
-    """Split arrays or matrices into sequential train and test subsets.
+    """Split time series data containers into a single train/test split.
 
-    Creates train/test splits over endogenous arrays an optional exogenous
-    arrays.
+    Splits time series ``y`` into a single temporally ordered train and test split.
+    The split is based on ``test_size`` and ``train_size`` parameters,
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply sktime/split/temporal_train_test_split.py:35]: that behaviour is still exactly as before.

The pre-PR code was branching into completely disjoint branches, one for `fh` being passed (the first "if" statement with return), one branch for `fh=None`.

This PR changes only the second branch, and leaves the first branch unchanged. A similar PR could fix the first branch.

File: sktime/split/temporal_train_test_split.py (hunk 2 above)
Comment (fkiraly) [reply sktime/split/temporal_train_test_split.py:103]: I thought the else block made the condition easier to spot for a reader of the code, because it matches the indentation and can be tracked up to the `if` condition.

If we remove the `else`, the `if` condition will be harder to find.

This is my rationale (in general) to do this even if you have `if`-`return`, but I'm not too strongly opined about this.

Comment (fkiraly) [reply sktime/split/temporal_train_test_split.py:103]: For instance, I did the opposite with `if fh is not None:`, because the `else` was very long.

Comment (fkiraly) [reply sktime/split/temporal_train_test_split.py]: I made a change to the code flow, and added comments - better? I went with your suggestion, but added comments to the `fh` and the `X` branching to make it clear to the reader

File: sktime/split/temporal_train_test_split.py [hunk 4]
```diff
@@ -83,46 +80,155 @@ def temporal_train_test_split(
...
+        If float, must be between 0.0 and 1.0, and is interpreted as the proportion
+        of the dataset to include in the test split. Proportions are rounded to the
+        next integer count of samples.
+        If int, is interpreted as total number of test samples.
+        If None, the value is set to the complement of the train size.
+        If ``train_size`` is also None, it will be set to 0.25.
+    train_size : float, int, or None, (default=None)
+        If float, must be between 0.0 and 1.0, and is interpreted as the proportion
+        of the dataset to include in the train split. Proportions are rounded to the
+        next integer count of samples.
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply sktime/split/temporal_train_test_split.py:151]: yes, fixed

File: sktime/split/temporal_train_test_split.py (hunk 2 above)
Comment (yarnabrina) [reply sktime/split/temporal_train_test_split.py]: It was a non-blocking comment really. Usually tools like `pylint` will complain against these (ref. [R1705](https://pylint.pycqa.org/en/stable/user_guide/messages/refactor/no-else-return.html)).

The change looks good to me. `pylint` or etc. may still complain that function return signature vary between different return statements (ref. [R1710](https://pylint.readthedocs.io/en/stable/user_guide/messages/refactor/inconsistent-return-statements.html)), but I don't think there's a way to solve that here at all.

**Review by yarnabrina — APPROVED** [review]
